### PR TITLE
Add automatic coloring of tags

### LINKED ISSUE #51: Automatic tag and correspondent colors
Inspired by what for example Grafana does (and due to the limited set of colors available in paperless), I patched it to auto-color tags and correspondents with a https://github.com/zenozeng/color-hash like implementation:

- https://github.com/jayme-github/paperless/commit/db0390cff12efe6bdb5d53d75a6b46d2016f0953
- https://github.com/jayme-github/paperless/commit/9def166811a0caec3adfe0c7b481390e174b462d

I understand that people might like choosing the colors themselves, so feel free to close this. :smile: It's just that I don't care about the actual color of tags, I just want them to have different ones and I don't want to bothered with selecting them.

### PR DESCRIPTION
**Pull request by jayme-github**
Please see this as proposal on how to implement automatic/random colors
for tags while keeping the current set of hard coded colors in place (at
least in the frontend).

Bear with me as I have even less Angular knowledge than Django and just
tried to get away with as few changes as possible. :-) AIUI you want to
change to a color picking system anyways in the future, which could also
play well with this.

fixes #51

### CODE REVIEW

**Comment by jonaswinkler** [comment]
Hello, and thanks again for contributing! (Don't worry too much about the coverage report)

The next release is already kinda packed with goodies, and I want to get this wrapped up, so I'll look into this alter.

**Comment by jayme-github** [comment]
> Hello, and thanks again for contributing! (Don't worry too much about the coverage report)
> 
> The next release is already kinda packed with goodies, and I want to get this wrapped up, so I'll look into this alter.

Sure, no rush. Let me know if you want anything changed.

File: src-ui/src/app/components/manage/tag-list/tag-list.component.ts [hunk 1]
```diff
@@ -15,10 +15,14 @@ export class TagListComponent extends GenericListComponent<PaperlessTag> {
 
   constructor(tagService: TagService, modalService: NgbModal) {
     super(tagService, modalService, TagEditDialogComponent)
-   }
+  }
 
   getColor(id) {
-    return TAG_COLOURS.find(c => c.id == id)
+    var color = TAG_COLOURS.find(c => c.id == id)
```
Comment (Tooa) [inline src-ui/src/app/components/manage/tag-list/tag-list.component.ts:21]: Is there a reason we mix US and British English spelling for color (colour)?

Comment (jayme-github) [reply src-ui/src/app/components/manage/tag-list/tag-list.component.ts:21]: No. I'm happy to clean that up once/if the general approach is accepted.

**Comment by jayme-github** [comment]
@jonaswinkler did you had the time to think about this by chance? Would you accept the approach in general? I could rebase and fix the spelling then.

**Comment by jonaswinkler** [comment]
@bauerj

I want to look into this again and need some input. The gist of it is that the field `Tag.colour` will not return an indexed color anymore, but rather a hex string `#ffabab`. (Fancy how github parses that!)

As far as I'm aware, your app is the only client that extensively uses the API. Are you using this field in your app? Would you mind me changing that? The more compatible approach would be to

* Add `colour_code` field
* Keep `colour` as a read only field that tries to return one of the predefined colors close to what's saved in `colour_code`.

**Comment by bauerj** [comment]
Yes, the app is currently using that field. Your workaround seems useful although it probably needs a lot of code to find the closest original color.

I guess  `colour` could just return a fixed number for non-default colors. This wouldn't break any important functionality and only for a very short period of time. I will implement an agreed-upon change quickly so we can have the app ready before users can use the new feature on the server-side.

App updates are usually installed within weeks at most so I guess you can drop `colour` entirely in a few versions.

For future changes it might be better to send an `X-API-Version` header or something. This could enable the app to prompt users to update if their server is too new.

**Comment by jonaswinkler** [comment]
> although it probably needs a lot of code to find the closest original color. I guess colour could just return a fixed number for non-default colors. 

True.

> For future changes it might be better to send an `X-API-Version` header or something. This could enable the app to prompt users to update if their server is too new.

Hm. I'll add API versioning. Time to learn something new! There's support for that in the framework, and I'm apparently able to alter the behavior based on what version is requested. Old clients get the old version, new clients get fancy color codes.

**Comment by bauerj** [comment]
Alright, I implemented a client-side change assuming the JSON key will be named "colour_code". It still works for the old API so I'm going to push an update.

I wonder if there is an easier way than to maintain backwards-compatibility client-side as well as server-side 😄

**Comment by jonaswinkler** [comment]
> Alright, I implemented a client-side change assuming the JSON key will be named "colour_code". It still works for the old API so I'm going to push an update.
> 
> I wonder if there is an easier way than to maintain backwards-compatibility client-side as well as server-side

Wait.

**Comment by bauerj** [comment]
Okay 😄

I ran into a new kind of [dependency hell](https://dart.dev/null-safety/migration-guide) anyway. Is there anything else that I would need to change?

**Comment by jonaswinkler** [comment]
I am about to implement versioning according to the documentation (https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning). This will ensure backward compatibility of all (older) clients with all servers.

* The default would be to serve v1 (which is the current state, and will serve compatible `colour` fields).
* If the version is specified, the behavior of the server will change according to the requested version.
* v2 and onward would essentially be v1, but with the changes to the Tag model. (`color_code` present, `colour` removed)

**Comment by bauerj** [comment]
Looks like the app has to maintain backwards-compatibility for old servers and the server has to maintain backwards-compatibility for old clients then 😅

I wonder what happens if the client requests an API version 3. Would the server just use the v2 API or would it raise an error?

> Adding that versioned Accept header to every request is fairly simple in the front end - how's that on your end?

Yes, that would be simple as well.

**Comment by jonaswinkler** [comment]
> I wonder what happens if the client requests an API version 3. Would the server just use the v2 API or would it raise an error?

Newer servers will respond with "406 Not Acceptable", and an error message in the body.

> Looks like the app has to maintain backwards-compatibility for old servers and the server has to maintain backwards-compatibility for old clients then

Not at all! You simply build your app to target one of the versions, and expect the server to respond with data from that version. The apps that are currently running out there will continue to request v1. When you update your app to use the new color fields, simply request version 2, and the server will serve data from that version, even if newer API versions appear in the future.

**Comment by jonaswinkler** [comment]
image

This took way too much time.

**Comment by shamoon** [comment]
> This took way too much time.

Heh, on the plus side it looks dope.

**Comment by jonaswinkler** [comment]
> For future changes it might be better to send an `X-API-Version` header or something. This could enable the app to prompt users to update if their server is too new.

With this you could verify whether your app is compatible with any given server, right? `X-API-Version` not present -> not compatible, `X-API-Version` present -> compatible if this version is high enough. Correct?

**Comment by bauerj** [comment]
>> Looks like the app has to maintain backwards-compatibility for old servers and the server has to maintain backwards-compatibility for old clients then

>Not at all! You simply build your app to target one of the versions, and expect the server to respond with data from that version. The apps that are currently running out there will continue to request v1. When you update your app to use the new color fields, simply request version 2, and the server will serve data from that version, even if newer API versions appear in the future.

Well, for the time being we still need to support Paperless and Paperless NG API version 1 servers. I don't think it would be a good idea to force users to upgrade their server after an automatic app update. 

While app updates are usually installed by almost all users after 1-2 days, server updates can take a lot longer since there is no integrated update mechanism.

However, this versioning scheme is still useful for other clients.

>>I wonder what happens if the client requests an API version 3. Would the server just use the v2 API or would it raise an error?

>Newer servers will respond with "406 Not Acceptable", and an error message in the body.

I'm not sure if this is a good idea. Imagine we have API version 10 some time in the future but the user would still use an old server that only supports version 2 of the API. Then the app would have to request data using API versions 10-2, causing 8 requests to find the common API version. Or does the 406 response contain an X-API-Version header with the newest API?

**Comment by jonaswinkler** [comment]
I'm not sure either how to do this properly.

Every response to *authenticated* requests will contain `X-Api-Version`.

**Comment by jonaswinkler** [comment]
What should I do when the client requests a version that the server does not (yet) have?

The current state of things on the `dev` branch is described in the documentation. https://paperless-ng.readthedocs.io/en/dev/api.html#api-versioning

**Pull request merged**